ambry-store/src/main/java/com.github.ambry.store/StoreDescriptor.java

```diff
@@ -45,26 +51,40 @@
   public StoreDescriptor(String dataDir) throws IOException {
     File storeDescriptorFile = new File(dataDir, STORE_DESCRIPTOR_FILENAME);
     if (storeDescriptorFile.exists()) {
-      DataInputStream dataInputStream = new DataInputStream(new FileInputStream(storeDescriptorFile));
-      short version = dataInputStream.readShort();
+      CrcInputStream crcStream = new CrcInputStream(new FileInputStream(storeDescriptorFile));
+      DataInputStream stream = new DataInputStream(crcStream);
+      short version = stream.readShort();
       switch (version) {
         case VERSION_0:
           // read incarnationId
-          String incarnationIdStr = Utils.readIntString(dataInputStream);
+          String incarnationIdStr = Utils.readIntString(stream);
           incarnationId = UUID.fromString(incarnationIdStr);
+          long crc = crcStream.getValue();
+          long crcValueInFile = stream.readLong();
+          if (crc != crcValueInFile) {
+            throw new IllegalStateException(
+                "CRC mismatch for StoreDescriptor. CRC of the stream " + crc + ", CRC from file " +
+                    crcValueInFile);
+          }
           break;
         default:
           throw new IllegalArgumentException("Unrecognized version in StoreDescriptor: " + version);
       }
     } else {
       incarnationId = UUID.randomUUID();
-      if (storeDescriptorFile.createNewFile()) {
-        DataOutputStream dataOutputStream = new DataOutputStream(new FileOutputStream(storeDescriptorFile));
-        dataOutputStream.write(toBytes());
-        dataOutputStream.close();
+      File tempFile = new File(dataDir, STORE_DESCRIPTOR_FILENAME + ".tmp");
+      File actual = new File(dataDir, STORE_DESCRIPTOR_FILENAME);
+      if (tempFile.createNewFile()) {
+        FileOutputStream fileStream = new FileOutputStream(tempFile);
+        CrcOutputStream crc = new CrcOutputStream(fileStream);
+        DataOutputStream writer = new DataOutputStream(crc);
+        writer.write(toBytes());
+        long crcValue = crc.getValue();
+        writer.writeLong(crcValue);
+        fileStream.getChannel().force(true);
+        tempFile.renameTo(actual);
```

Comment:
this operation can fail (it returns `true` or `false`).

Intellij will be indicating this things in yellow - please take a look whenever there are intellij hints - they are usually useful.